stablehlo/reference/Element.cpp

```diff
@@ -720,6 +720,57 @@ Element rsqrt(const Element &el) {
       [](std::complex<double> e) { return 1.0 / std::sqrt(e); });
 }
 
+Element sign(const Element &el) {
```

Comment:
Sorry for the repeated reviews! A few more issues popped up after enabling the CI. Here is the comprehensive fixes:
```
Element sign(const Element &el) {
  Type type = el.getType();

  if (isSupportedIntegerType(type)) {
    auto intEl = el.getIntegerValue();

    auto isNonZero = Element(type, (int64_t)!intEl.isZero());
    return intEl.isNegative() ? -isNonZero : isNonZero;
  }

  if (isSupportedFloatType(type)) {
    auto elVal = el.getFloatValue();

    if (elVal.isNaN()) return el;
    auto isNonZero = Element(type, (double)!elVal.isZero());
    return elVal.isNegative() ? -isNonZero : isNonZero;
  }

  if (isSupportedComplexType(type)) {
    auto elVal = el.getComplexValue();
    const llvm::fltSemantics &elSemantics = elVal.real().getSemantics();

    if (elVal.real().isNaN() || elVal.imag().isNaN())
      return Element(type, std::complex<APFloat>(APFloat::getNaN(elSemantics),
                                                 APFloat::getNaN(elSemantics)));

    auto absElVal = abs(el).getFloatValue();
    if (absElVal.isZero()) return Element(type, (double)0.0);

    bool roundingErr;
    APFloat resultReal(elVal.real().convertToDouble() /
                       absElVal.convertToDouble());
    resultReal.convert(elSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
    APFloat resultImag(elVal.imag().convertToDouble() /
                       absElVal.convertToDouble());
    resultImag.convert(elSemantics, APFloat::rmNearestTiesToEven, &roundingErr);

    return Element(type, std::complex<APFloat>(resultReal, resultImag));
  }

  report_fatal_error(invalidArgument("Unsupported element type: %s",
                                     debugString(type).c_str()));
}
```
Feedback:
* Reduced logic in handling integer and float type.
* Unnecessary wrapper around `el`: `if (elVal.isNaN()) return Element(el);` -> `if (elVal.isNaN()) return el;`.
* HLO returns NaNs for both real and imaginary parts, which makes sense since `abs(NaN) = NaN` and divide by NaN should also be NaN. We should also follow that.

Reply:
"Reduced logic in handling integer and float type". I would propose that we aim to have 1:1 correspondence with the code provided in the spec. 

The implementation proposed in the comment is clever, but a non-zero mental effort is needed to correlate it with the spec. Instead, how about this (I slightly reordered the implementation provided in the current version of the PR):

```C++
  if (isSupportedIntegerType(type)) {
    auto elVal = el.getIntegerValue();
    if (intEl.isNegative()) return Element(type, (int64_t)-1);
    if (intEl.isZero()) return Element(type, (int64_t)0);
    return Element(type, (int64_t)1);
  }

  if (isSupportedFloatType(type)) {
    auto elVal = el.getFloatValue();
    if (elVal.isNaN()) return Element(el);
    if (elVal.isNegative()) return Element(type, (double)-1.0);
    if (elVal.isNegZero()) return Element(type, (double)-0.0);
    if (elVal.isPosZero()) return Element(type, (double)+0.0);
    return Element(type, (double)1.0);
  }
```

Reply:
Hello @burmako ,

The logic for floating-type has an error here.

`elVal.isNegative()` will be `true` for `-0.0` as well.
Then it will then return `-1.0` instead of `-0.0`.

The order should be the following imo:

```c++
  if (isSupportedFloatType(type)) {
    auto elVal = el.getFloatValue();
    if (elVal.isNaN()) return el;
    if (elVal.isNegZero()) return Element(type, (double)-0.0);
    if (elVal.isPosZero()) return Element(type, (double)+0.0);
    if (elVal.isNegative()) return Element(type, (double)-1.0);
    return Element(type, (double)1.0);
  }
```

Reply:
Nice catch! I've updated #1395 to reflect this order as well.

Reply:
Thanks 😄